File: src/main/java/de/jepfa/obfusser/ui/common/Debug.java
package de.jepfa.obfusser.ui.common;

import android.app.Activity;
import android.graphics.drawable.Drawable;
import android.support.v7.app.AlertDialog;

import de.jepfa.obfusser.BuildConfig;
import de.jepfa.obfusser.Constants;
import de.jepfa.obfusser.model.Secret;
import de.jepfa.obfusser.ui.SecureActivity;
import de.jepfa.obfusser.util.EncryptUtil;

public class Debug {

    private static boolean debug = false;

    public static void showDebugDialog(Activity activity) {
        AlertDialog.Builder builder = new AlertDialog.Builder(activity);

        StringBuilder sb = new StringBuilder();

        byte[] salt = SecureActivity.SecretChecker.getSalt(activity);
        addParam(sb, "AppSalt", endOfArrayToString(salt, 4) + ", l=" + arrayLength(salt));

        byte[] key = Secret.getOrCreate().getDigest();
        addParam(sb, "Key", endOfArrayToString(key, 4) + ", l=" + arrayLength(key));

        addParam(sb, "Key outdated", String.valueOf(Secret.getOrCreate().isOutdated()));

        addParam(sb, "Enc supported", String.valueOf(EncryptUtil.isPasswdEncryptionSupported()));
        addParam(sb, "Key stored", String.valueOf(SecureActivity.SecretChecker.isPasswordStored(activity)));
        addParam(sb, "Salt encrypted", String.valueOf(SecureActivity.SecretChecker.isSaltEncrypted(activity)));
        addParam(sb, "Enc with UUID", String.valueOf(SecureActivity.SecretChecker.isEncWithUUIDEnabled(activity)));

        Drawable icon = activity.getApplicationInfo().loadIcon(activity.getPackageManager());
        builder.setTitle("Debug info")
                .setMessage(sb.toString())
                .setIcon(icon)
                .show();
    }

    public static synchronized void toggleDebug() {
        Debug.debug = !Debug.debug;
    }

    public static boolean isDebug() {
        return BuildConfig.DEBUG || debug == true;
    }

    private static void addParam(StringBuilder sb, String name, String value) {
        sb.append(name);
        sb.append(" = ");
        sb.append(value);
        sb.append(Constants.NL);
    }

    public static String endOfArrayToString(byte[] a, int count) {
        if (a == null)
            return "null";
        int iMin = Math.max(a.length - count, 0);
        int iMax = a.length - 1;
        if (iMax == -1)
            return "..]";

        StringBuilder b = new StringBuilder();
        b.append(".., ");
        for (int i = iMin; ; i++) {
            b.append(a[i]);
            if (i == iMax)
                return b.append(']').toString();
            b.append(", ");
        }
    }

    public static String arrayLength(byte[] a) {
        if (a == null) {
            return "n/a";
        }
        return String.valueOf(a.length);
    }
}

File: src/main/java/de/jepfa/obfusser/ui/common/Debug.kt
package de.jepfa.obfusser.ui.common

import android.app.Activity
import android.os.Build
import android.support.v7.app.AlertDialog
import android.util.Log

import de.jepfa.obfusser.BuildConfig
import de.jepfa.obfusser.Constants
import de.jepfa.obfusser.model.Secret
import de.jepfa.obfusser.ui.SecureActivity
import de.jepfa.obfusser.util.encrypt.EncryptUtil

object Debug {

    private var debug = BuildConfig.DEBUG

    val isDebug: Boolean
        get() = debug

    fun showDebugDialog(activity: Activity) {
        val builder = AlertDialog.Builder(activity)

        val sb = StringBuilder()

        try {
            val pInfo = activity.application.packageManager.getPackageInfo(activity.application.packageName, 0)
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                addParam(sb, "VersionCode", pInfo.longVersionCode.toString())
            }
            else {
                addParam(sb, "VersionCode", pInfo.versionCode.toString())
            }
            addParam(sb, "VersionName", pInfo.versionName)
        } catch (e: Exception) {
            Log.e("DEBUGINFO", "cannot get version code", e)
        }

        addParam(sb, "BuildTimestamp", Constants.SDF_DT_MEDIUM.format(BuildConfig.BUILD_TIME))
        addParam(sb, "BuildType", BuildConfig.BUILD_TYPE)
        addParam(sb, "SdkVersion", Build.VERSION.SDK_INT.toString())

        val salt = SecureActivity.SecretChecker.getSalt(activity)
        addParam(sb, "AppSalt", byteArrayChecksum(salt).toString() + ", len=" + arrayLength(salt))

        val key = Secret.getOrCreate().digest
        addParam(sb, "Key", byteArrayChecksum(key).toString() + ", len=" + arrayLength(key))

        addParam(sb, "Key outdated", Secret.getOrCreate().isOutdated.toString())

        addParam(sb, "Enc supported", EncryptUtil.isPasswdEncryptionSupported().toString())
        addParam(sb, "Key stored", SecureActivity.SecretChecker.isPasswordStored(activity).toString())
        addParam(sb, "Salt encrypted", SecureActivity.SecretChecker.isSaltEncrypted(activity).toString())
        addParam(sb, "Enc with UUID", SecureActivity.SecretChecker.isEncWithUUIDEnabled(activity).toString())
        addParam(sb, "Uncrypted Strings", SecureActivity.SecretChecker.shouldDoCryptStrings(activity).toString())

        val icon = activity.applicationInfo.loadIcon(activity.packageManager)
        builder.setTitle("Debug info")
                .setMessage(sb.toString())
                .setIcon(icon)
                .show()
    }

    @Synchronized
    fun toggleDebug() {
        Debug.debug = !Debug.debug
    }

    private fun addParam(sb: StringBuilder, name: String, value: String) {
        sb.append(name)
        sb.append(" = ")
        sb.append(value)
        sb.append(Constants.NL)
    }

    fun byteArrayChecksum(bytes: ByteArray?): Int {
        if (bytes == null)
            return 0

        var checkSum = 0
        for (b in bytes) {
            checkSum += b
        }
        return checkSum
    }

    private fun arrayLength(a: ByteArray?): String {
        return a?.size?.toString() ?: "n/a"
    }
}
